### PR TITLE
[RLlib][Experimental][WIP] Torch mixed precision in rllib

### PR DESCRIPTION
**Pull request by smorad**
This is experimental use of `torch.amp` mixed-precision with PPO (I expect this to fail with other algorithms at this point). Not too many changes are required to autocast all torch operations to `bfloat16`, which supposedly produces a speedup of 4x on CUDA tensor cores.

My laptop does not have a GPU so I can't perf test (CPU `bfloat16` will not produce any perf benefits). Once I get access to a GPU I can get some perf numbers. Surprisingly, we get very similar model accuracy on cartpole, even without a gradient scaler. I suspect this is because the gradient noise from taking the expectation over rollouts dwarves whatever noise is produced from the lost precision. The following results are `torch.amp/bfloat16` (blue) and normal master `fp32` (orange).
Screen Shot 2022-02-14 at 4 17 41 PM

## Related issue number

Following investigation laid out in #22166

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: rllib/models/torch/fcnet.py [hunk 1]
```diff
@@ -142,11 +142,16 @@ def forward(
         seq_lens: TensorType,
     ) -> (TensorType, List[TensorType]):
         obs = input_dict["obs_flat"].float()
-        self._last_flat_in = obs.reshape(obs.shape[0], -1)
-        self._features = self._hidden_layers(self._last_flat_in)
-        logits = self._logits(self._features) if self._logits else self._features
-        if self.free_log_std:
-            logits = self._append_free_log_std(logits)
+        # TODO: use the correct type for CUDA
+        with torch.autocast(
```
Comment (gjoliver) [inline rllib/models/torch/fcnet.py:146]: the torch version we use in CI doesn't have autocast, it seems.

Comment (smorad) [reply rllib/models/torch/fcnet.py:146]: Oops! It should. I think in `pytorch-1.10` they alias `torch.cuda.amp.autocast` to `torch.autocast`. I will use the longer version instead. I don't think in the older versions in works on CPU, but it's kinda pointless to use it on CPU anyways.

Comment (smorad) [reply rllib/models/torch/fcnet.py:146]: For `torch-1.9` as listed in https://github.com/ray-project/ray/blob/53c4c7b1be54fe67a54b34f3f0c076ad502e35cc/python/requirements_ml_docker.txt#L14-L15: https://pytorch.org/docs/1.9.0/amp.html#torch.cuda.amp.autocast

Comment (gjoliver) [reply rllib/models/torch/fcnet.py:146]: cool. let's see if the tests pass after it.

also a really clueless question, why is this controlled by obs.is_cuda bit, and not say, the actual nn model?

Comment (smorad) [reply rllib/models/torch/fcnet.py:146]: So devices in torch are only defined for `Tensors`. `torch.nn.Modules` do not have a device. This is because maybe you want some of your layers on the CPU and some on the GPU. It can be annoying, because often your model is only on one device, but that's the price of flexibility I suppose.

Instead, we can just check what device the tensors are on. If we try and multiply `obs.cuda() * torch.tensor(1.).cuda()` we'll get an error that `obs` and `1` are on different devices. So we can be pretty sure that the `obs` device is the same as all the parameters in the model.

Comment (gjoliver) [reply rllib/models/torch/fcnet.py:146]: interesting. thanks for the explanation.

**Comment by smorad** [comment]
Comment to postpone stalebot
